src/backend/api/handlers.py

```diff
@@ -0,0 +1,191 @@
+
+
+def _get_distance_between(
+        lat1: float,
+        lon1: float,
+        lat2: float,
+        lon2: float,
+) -> float:
+    return math.sqrt(
+        (lat1 - lat2)**2 + (lon1 - lon2)**2
+    )
+
+
+def _get_nearest_mirrors(ip_address):
+    db = GeoIPEngine.get_instance()
+    match = db.lookup(ip_address)
+    with session_scope() as session:
+        if match is None:
+            all_mirrors = session.query(Mirror).filter(
+                Mirror.is_expired == false(),
+                ).all()
+            all_mirrors = [mirror.to_dict() for mirror in all_mirrors]
+            return all_mirrors
+        match_dict = match.get_info_dict()
+        match_country = match_dict['country']['names']['en']
+        match_continent = match_dict['continent']['names']['en']
+        suitable_mirrors = session.query(Mirror).filter(
+            Mirror.continent == match_continent,
+            Mirror.country == match_country,
+            Mirror.is_expired == false(),
+        ).limit(MAX_LENGTH_OF_MIRRORS_LIST).all()
+        suitable_mirrors = [mirror.to_dict() for mirror in suitable_mirrors]
+        if len(suitable_mirrors) == MAX_LENGTH_OF_MIRRORS_LIST:
+            return suitable_mirrors
+        rest_mirrors = session.query(Mirror).filter(
+            Mirror.continent == match_continent,
+            Mirror.is_expired == false(),
+            Mirror.name.notin_(
+                [mirror.name for mirror in suitable_mirrors]
+            ),
+        ).all()
+        rest_mirrors = [mirror.to_dict() for mirror in rest_mirrors]
+        if not rest_mirrors:
+            rest_mirrors = session.query(Mirror).filter(
+                Mirror.name.notin_(
+                    [mirror.name for mirror in suitable_mirrors]
+                ),
+                Mirror.is_expired == false(),
+            ).all()
+            rest_mirrors = [mirror.to_dict() for mirror in rest_mirrors]
+        distance_dict = {
+            _get_distance_between(
+                lat1=match.location[0],
+                lon1=match.location[1],
+                lat2=mirror['location']['lat'],
+                lon2=mirror['location']['lon'],
+            ): mirror for mirror in rest_mirrors
+        }
+        suitable_mirrors.extend(
+            [distance_dict[distance] for distance in
+             sorted(distance_dict.keys())
+             [:MAX_LENGTH_OF_MIRRORS_LIST - len(suitable_mirrors)]]
+        )
+        return suitable_mirrors
+
+
+def update_mirrors_handler():
+    config = get_config()
+    versions = config['version']
+    repos = config['repos']
+    verified_mirrors = get_verified_mirrors(
+        mirrors_dir=os.path.join(
+            os.path.dirname(
+                os.path.abspath(__file__),
+            ),
+            '../../../mirrors',
+            config['mirrors_dir'],
+        ),
+        versions=versions,
+        repos=repos,
+        allowed_outdate=config['allowed_outdate']
+    )
+    with session_scope() as session:
```

Comment:
Does session_scope apply atomic() inside? Because otherwise we may remove all mirrors and don't get them back.

Reply:
Yes, I'm applying a transaction after all operations. And I disabled autoflush.